### ISSUE TITLE
Environment variables should be propagated to workers from the driver

### ISSUE THREAD

**Issue by ericl** [ray-project/ray#2260]
### Describe the problem
Enironment variables aren't propagated to the worker processes in cluster mode.

### Source code / logs
```
import ray
import os

ray.init(redis_address="localhost:6379")

@ray.remote
def get_foo():
    return os.environ["FOO"]

print(ray.get(get_foo.remote()))
```

If you run `FOO=1 python test.py` as above, the remote task will crash with `KeyError: 'FOO'`. However it will run if you use a bare `ray.init()`.

A workaround is to set the env vars when starting the cluster, e.g., `FOO=1 ray start`.

**Comment by robertnishihara** [comment]
The environment variables on the workers are inherited from the local scheduler or raylet process that creates them, which in turn is created by the `ray start` command (or the `ray.init()` command), so if you set the environment variable before calling `ray start` (on each machine), then it will work (or before starting the Python process if you are calling `ray.init()`.

Whenever a worker executes a task, we could in theory swap in the environment variables for that task's job and then swap them back out when the next task runs. However, a lot of these environment variables only matter at certain points in time (e.g., setting `MKL_NUM_THREADS` doesn't help after importing numpy I think, and the workers may have imported numpy long before the job is executed).

So for this proposal to make sense, we probably also need to not share workers between different jobs (which is probably a good idea anyway).

**Comment by thebeancounter** [comment]
@robertnishihara 
is there any way to give a worker a specific value for env variable? 
e.g give one worker value of 1 and another value of 2 ?

thanks

**Comment by richardliaw** [comment]
You have to manually do that.

**Comment by thebeancounter** [comment]
@richardliaw how? in the machine itself? in ray?

**Comment by frthjf** [comment]
I'm running into this issue where the environment variable has to be available before the worker imports a certain package. 
> A workaround is to set the env vars when starting the cluster, e.g., FOO=1 ray start.

Would it perhaps be possible to leverage that workaround in the autoscaler config, e.g.
```yaml
worker_start_ray_commands:
    - ray stop
    - ${OS_ENV_STRING HERE} ray start --address=$RAY_HEAD_IP:6379
```

**Comment by richardliaw** [comment]
Yeah, that should work.

**Comment by frthjf** [comment]
Thanks @richardliaw, I think I can use that as a workaround. How hard would it be to fix this issue, i.e. (1) not share workers between different jobs and (2) swap in the environment variables for tasks as @robertnishihara suggested? I'd be willing to contribute if it's not a too difficult undertaking.

**Comment by richardliaw** [comment]
You can set “max_calls=1” in ray.remote to disable worker sharing. What API
would you like for (2)?

**Comment by frthjf** [comment]
Great, so (1) is already in place. For (2), it looks like it could be handled automatically in the .remote() call ```print(ray.get(get_foo.remote()))```. It would need to capture available env-vars and send them to the worker who swaps them in before execution.

However, I'm not sure how useful that actually is because I believe import statements will have already happened at this point:
```python
import ray
import os

os.environ["MKL_NUM_THREADS"] = "1"  # env-var that acts on import below
import numpy as np

ray.init(redis_address="localhost:6379")

@ray.remote
def get_foo():
    # even with env-vars being swapped in 
    # the numpy instance will ignore MKL_NUM_THREADS right?
    return np.zeros([1])  

print(ray.get(get_foo.remote()))
```
Although I'm not 100% sure. Another issue are environment variable conflicts, like PATH that is likely to be set on the driver and the remote worker.
